File: android/src/main/java/expo/modules/nodemediaclient/ExpoNodePlayerView.kt
package expo.modules.nodemediaclient

import android.content.Context
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.viewevent.EventDispatcher
import expo.modules.kotlin.views.ExpoView

import android.util.Log
import android.widget.FrameLayout
import cn.nodemedia.NodePlayer

class ExpoNodePlayerView(context: Context, appContext: AppContext) : ExpoView(context, appContext) {
    private val TAG = "ExpoNodePlayerView"
    private var np: NodePlayer? = null
    private val onEventCallback by EventDispatcher()

    var url = ""
    var cryptoKey = "" // Add this property
        set(value) {
            field = value
            np?.setCryptoKey(value)
        }
    var HTTPReferer = ""
        set(value) {
            field = value
            np?.setHTTPReferer(value)
        }
    var HTTPUserAgent = ""
        set(value) {
            field = value
            np?.setHTTPUserAgent(value)
        }
    var volume = 1.0f
        set(value) {
            field = value
            np?.setVolume(value)
        }
    var scaleMode = 0
        set(value) {
            field = value
            np?.setScaleMode(value)
        }
    var bufferTime = 1000
        set(value) {
            field = value
            Log.d(TAG, "bufferTime set to $value ms $np")
            np?.setBufferTime(value)
        }
    var RTSPTransport = "udp"
        set(value) {
            field = value
            np?.setRTSPTransport(value)
        }

    var HWAccelEnable = true
        set(value) {
            field = value
            np?.setHWAccelEnable(value)
        }
    private val videoView = FrameLayout(context).apply {
        layoutParams = LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT)
    }

    init {
        addView(videoView)
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        np = NodePlayer(context, LICENSE_KEY)
        np?.setOnNodePlayerEventListener { obj, event, msg ->
            onEventCallback(mapOf("event" to event, "msg" to msg))
        }
        np?.attachView(videoView)
    }

    override fun onDetachedFromWindow() {
        np?.stop()
        np?.detachView()
        np = null
        super.onDetachedFromWindow()
    }

    fun start(url: String?) {
        if (!url.isNullOrEmpty()) {
            this.url = url
        }
        np?.setVolume(this.volume)
        np?.setCryptoKey(this.cryptoKey)
        np?.setScaleMode(this.scaleMode)
        np?.setBufferTime(this.bufferTime)
        np?.setHTTPReferer(this.HTTPReferer)
        np?.setHTTPUserAgent(this.HTTPUserAgent)
        np?.setRTSPTransport(this.RTSPTransport)
        np?.setHWAccelEnable(this.HWAccelEnable)
        np?.start(this.url)
    }

    fun stop() {
        np?.stop()
    }
}

File: ios/ExpoNodePlayerView.swift
//
//  Copyright (c) 2025 NodeMedia Technology Co., Ltd.
//  Created by Chen Mingliang on 2025-07-23.
//  All rights reserved.
//

import ExpoModulesCore
import NodeMediaClient

class ExpoNodePlayerView: ExpoView, NodePlayerDelegate {
    private var _np: NodePlayer?
    var url = ""
    var croptyKey = ""
    var HTTPReferer = ""
    var HTTPUserAgent = ""
    var RTSPTransport = "udp"
    var HWAccelEnable = true
    private var _volume: Float = 1.0
    private var _scaleMode: Int = 0
    private var _bufferTime: Int = 1000

    // EventDispatcher for sending events to JavaScript
    var onEventCallback = EventDispatcher()

    var volume:Float {
        get {
            return _volume
        }
        set {
            _volume = newValue
            _np?.volume = newValue
        }
    }
    
    var scaleMode:Int {
        get {
            return _scaleMode
        }
        set {
            _scaleMode = newValue
            _np?.scaleMode = newValue
        }
    }
    
    var bufferTime:Int {
        get {
            return _bufferTime
        }
        set {
            _bufferTime = newValue
            _np?.bufferTime = newValue
        }
    }
    
    required init(appContext: AppContext? = nil) {
        super.init(appContext: appContext)
        _np = NodePlayer(license: LICENSE_KEY)
        _np?.nodePlayerDelegate = self
        _np?.attach(self)
    }
    
    deinit {
        _np?.stop()
        _np?.detachView()
        _np = nil
    }
    
    func start(u: String?) {
        if let newUrl = u {
            url = newUrl
        }
        _np?.volume = volume
        _np?.cryptoKey = croptyKey
        _np?.scaleMode = scaleMode
        _np?.bufferTime = bufferTime
        _np?.httpReferer = HTTPReferer
        _np?.httpUserAgent = HTTPUserAgent
        _np?.rtspTransport = RTSPTransport
        _np?.hwAccelEnable = HWAccelEnable
        _np?.start(url)
    }
    
    func stop() {
        _np?.stop()
    }
    
    func onEventCallback(_ sender: Any, event: Int32, msg: String) {
        onEventCallback([
            "event": Int(event),
            "msg": msg
        ])
    }
    
}
